### DISCUSSION TITLE
Broad and Narrow Phase

### DISCUSSION THREAD

**Discussion by Wishyou-D** [bepu/bepuphysics2#272] [Q&A]
Hey, I was wondering what functions are the start and end of broad and narrow phase? How do you define these two phases? Thanks.

**Comment by RossNordby** [comment]
The "broad phase" covers the updating and use of the simulation-wide acceleration structure. It's split into an update and testing phase: the update phase refits and refines the acceleration structure for the simulation's current state, and the testing phase finds collision pair candidates to feed to the narrow phase.

The narrow phase covers the detailed testing of those collision pair candidates and the bookkeeping associated with the results. The bookkeeping involves updating narrowphase internal caches (like the set of pairs), getting rid of stale collision constraints, adding new collision constraints, and waking up sleeping bodies affected by new constraints.

You can see how the broad phase are narrow phase are used in the [`Simulation.CollisionDetection` function](https://github.com/bepu/bepuphysics2/blob/master/BepuPhysics/Simulation.cs#L221).

The broad phase is updated for the current simulation state, then pair finding tests are invoked (which dumps its results into the narrow phase), and then the narrow phase flush is invoked to handle all the necessary bookkeeping.
